gax-java/gax/src/main/java/com/google/api/gax/rpc/StubSettings.java

```diff
@@ -99,17 +96,35 @@ protected StubSettings(Builder builder) {
     this.headerProvider = builder.headerProvider;
     this.internalHeaderProvider = builder.internalHeaderProvider;
     this.clock = builder.clock;
-    this.serviceName = builder.serviceName;
-    this.endpoint = builder.endpoint;
-    this.mtlsEndpoint = builder.mtlsEndpoint;
     this.switchToMtlsEndpointAllowed = builder.switchToMtlsEndpointAllowed;
     this.quotaProjectId = builder.quotaProjectId;
     this.streamWatchdogProvider = builder.streamWatchdogProvider;
     this.streamWatchdogCheckInterval = builder.streamWatchdogCheckInterval;
     this.tracerFactory = builder.tracerFactory;
     this.deprecatedExecutorProviderSet = builder.deprecatedExecutorProviderSet;
     this.gdchApiAudience = builder.gdchApiAudience;
-    this.universeDomain = builder.universeDomain;
+    this.endpointContext = buildEndpointContext(builder);
+  }
+
+  /**
+   * Attempt to build the EndpointContext from the Builder based on all the user configurations
+   * passed in.
+   *
+   * @throws RuntimeException if there is an issue building the EndpointContext
+   */
+  private EndpointContext buildEndpointContext(Builder builder) {
+    try {
+      return EndpointContext.newBuilder()
+          .setServiceName(getServiceName())
```

Comment:
Pass the serviceName to the EndpointContext. For GAPICs, the child StubSettings should have a valid serviceName.